### PR TITLE
Fluent Bit Lambda Extensions Example added

### PR DESCRIPTION
**Pull request by youngjeong46**
*Issue #, if available:*

*Description of changes:*

Added extensions example for Fluent-Bit. Requesting Code Review.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-lambda-extension-fluent-bit/main.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+func main() {
+	extensionName := path.Base(os.Args[0])
+	printPrefix := fmt.Sprintf("[%s]", extensionName)
+	logger := log.WithFields(log.Fields{"agent": extensionName})
+
+	// Determines listener port, use default 1234 if not provided as Lambda environment variable
+	listenerPort := os.Getenv("PORT")
+	if os.Getenv("PORT") == "" {
+		listenerPort = defaultListenerPort
+	}
```
Comment (julianwood) [inline aws-lambda-extension-fluent-bit/main.go:58]: Would it be better to use another variable name rather than the generic PORT. All extensions and the function share ENVVARS so perhaps making it more specific to the extension with something such as FLB_PORT would be preferable

File: aws-lambda-extension-fluent-bit/main.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	"path"
+	"syscall"
+
+	extension "aws-lambda-extensions/aws-lambda-extensions-fluent-bit/extensions"
+	"aws-lambda-extensions/aws-lambda-extensions-fluent-bit/logsapi"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	defaultListenerPort = "1234" // DefaultListenerPort sets the http port to 1234 if not provided by function env variable "HOST"
```
Comment (julianwood) [inline aws-lambda-extension-fluent-bit/main.go:21]: This isn't the variable, should be "PORT" or changed based on other suggestion.

**Review by julianwood — COMMENTED** [review]
Thanks for the request, just a ENVVAR naming suggestion.

File: aws-lambda-extension-fluent-bit/main.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+		listenerPort = defaultListenerPort
+	}
+
+	// Resolves host address "sandbox" for Fluent Bit
+	addr, DNSErr := net.LookupHost("sandbox")
+	if DNSErr != nil {
+		logger.Fatal("The DNS name sandbox cannot be resolved.")
+	}
+	host := addr[0]
+	err := os.Setenv("HOST", host)
```
Comment (santiagocardenas) [inline aws-lambda-extension-fluent-bit/main.go:66]: `FLB_HOST`

File: aws-lambda-extension-fluent-bit/main.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	extension "aws-lambda-extensions/aws-lambda-extensions-fluent-bit/extensions"
+	"aws-lambda-extensions/aws-lambda-extensions-fluent-bit/logsapi"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	defaultListenerPort = "1234" // DefaultListenerPort sets the http port to 1234 if not provided by function env variable "HOST"
+	maxItems            = 10000
+	maxBytes            = 262144
+	timeOut             = 1000
```
Comment (santiagocardenas) [inline aws-lambda-extension-fluent-bit/main.go:24]: Reduce this

File: aws-lambda-extension-fluent-bit/README.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
+# Fluent Bit Lambda Extensions in Go
```
Comment (PettitWesley) [inline aws-lambda-extension-fluent-bit/README.md:1]: nit: Fluent Bit Logs Extension for Lambda

File: aws-lambda-extension-fluent-bit/README.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+2. Supported AWS Fluent Bit Outputs: CloudWatch Logs, Kinesis Firehose and Kinesis Data Streams.
+
+This extension runs a goroutine that will do the following:
+* Starts the Extension and registers to ExtensionAPI,
+* Invokes Fluent Bit binary using user-specified configurations,
+* Subscribes the extension to the Logs API
+
+## Prerequisites
+
+1. You need to compile a Fluent Bit binary. The binary should be placed inside the `fluent-bit/` directory. Here are the steps:
```
Comment (PettitWesley) [inline aws-lambda-extension-fluent-bit/README.md:16]: Can we include the binary you have been testing with in this repo? That way users can run the example without this step. You can reword it to say that its to update the binary to a new version.

File: aws-lambda-extension-fluent-bit/fluent-bit/fluent-bit.conf [hunk 7]
```diff
@@ -0,0 +1,23 @@
+[SERVICE]
+    Flush           1
+    Grace           0
+    Parsers_File    parsers.conf
+
+[INPUT]
+    Name            http
+    Tag             lambda.http
+    Listen          ${HOST}
+    Port            ${PORT}
```
Comment (PettitWesley) [inline aws-lambda-extension-fluent-bit/fluent-bit/fluent-bit.conf:10]: this seems to be missing the new `tag_key` field that I added

File: aws-lambda-extension-fluent-bit/README.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
+# Fluent Bit Lambda Extensions in Go
+
+The provided code sample demonstrates how to get Fluent Bit set up for Lambda Extensions. 
+
+Note: 
+1. Supported Fluent Bit: v1.6
```
Comment (PettitWesley) [inline aws-lambda-extension-fluent-bit/README.md:6]: This is not true. My code changes have not been merged yet, but once they are, it will mean Lambda is supported in Fluent Bit version 1.7.4 or greater

Comment (PettitWesley) [reply aws-lambda-extension-fluent-bit/README.md:6]: Also, link to the official Fluent Bit docs somewhere in the beginning

**Comment by r0mdau** [comment]
Thanks for this example.

Have you planned (or discussed with fluentbit maintainers) to develop directly a [input plugin](https://github.com/fluent/fluent-bit/tree/master/plugins/in_http) in fluentbit that implement the lambda logs API ?

**Comment by PettitWesley** [comment]
**NOTE**: For folks who find this and wonder why we abandoned it, it was because the extension tended to lose logs when the lambda function was invoked repeatedly with a very short runtime. It is possible that this solution can work well if you know that your function will always execute for at least 1 second.

@r0mdau We discussed building native support in FB but that would require a lot more work. It's very non-trivial.

**Comment by julianwood** [comment]
Closing based on submitter comments.

**Comment by RaJiska** [comment]
Hi @PettitWesley , @youngjeong46 , @julianwood , @r0mdau .

I stumbled upon this PR as I'm looking to have FB as part of Lambda as an extension. Would you be able to share more details about the issue and what you tried so far?

This may be obvious and you probably already considered, although I can't help but notice that @PettitWesley you mentioned logs would be lost for runtimes below 1s. Correlating with the [configuration file](https://github.com/aws-samples/aws-lambda-extensions/pull/42/changes#diff-95983028dd0d107e0f00d86f3aad2f5a6ede91d86698032b2f02fea3b9089eceR2-R3) in use I noticed you set  `grace` to `0` and `flush` to `1`, wouldn't the grace immediately shut down Fluentbit before flushing, and therefore result in the aforementioned issue?

External extensions are allowed up to 2000ms to shutdown as per [documentation](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtime-environment.html#runtimes-lifecycle-shutdown). As the `grace` setting is parsed and [treated as an integer](https://github.com/fluent/fluent-bit/blob/master/src/flb_config.c#L58-L60), a safe value would be exactly just `1` (`0` prevents flushing, `2` would be at odd of not shutting down gracefully before the 2s AWS timeout) assuming no change to Fluentbit. This would therefore require settings to be something like `flush` set to `0.5` and `grace` to `1`, leaving at least 500ms to flush. Alternatively `grace` can be set to `2` but would likely be SIGKILL'd (and therefore marking the extension as erroring in aws logs) if the entire buffer is not flushed within the first per-second grace tick.
